RFC-0017-PyTorch-Operator-Versioning.md

```diff
@@ -0,0 +1,223 @@
+
+[Example changes](#heading=h.enajdjh00ukn)
+        * Note: this will not cover functional operators
+        * The default value is zero
+    * **Use a single operator version number for all operators**
+        * This number may be shared by the deploy version, but separate from other file format versions
+    * **Older versions of operators registration are kept but they should only be matchable with special circumstances**
+    * **Newer version of the operator registry must specify an upgrader** that conforms to the older version of the operator schema. Its body is a TorchScript-able function that uses the newer version of operator to implement old semantics.
+    * [Improved BC testing] Tests that the old serialized version of the operator can still be loaded on the new runtime and run as expected need to be easy to add
+        * This seems straightforward for Torchscript and Edge testing but I’m not sure how it would work for deploy/package
+    * [Improved FC testing] Tests that the new version of the operator can still be loaded on old runtimes and run as expected need to be easy to add 
+        * This might require a new test job, which could be tricky to setup
+* **Torchscript changes**
+    * TorchScript Runtime build contains a table of operators and their corresponding versions
+    * Each serialized model contains a table of operators and their corresponding version according to the TorchScript compiler that generated the model
+        * Q: will this require a new serialization format?
+        * Yes, but it is easy to make it BC and FC
+    * During loading into the TorchScript compiler, TorchScript needs to match operator schema according to the table of operator versions stored in the package. This would generate IR that conforms to older schema. 
+    * Then a TorchScript pass takes IR of older schema and replaces older versions of operator invocations with bodies of upgraders.
+    * Out-of-support operator versions (ie. those no longer defined in `native_functions.yaml` with a valid upgrader) need to throw an error
+* **Edge runtime and mobile delivery service changes**
+    * Delivery compatibility: communicating operator version table deployed on device and deliver models appropriately
+    * Runtime: load upgraders at model loading time, to ensure older models always work after updating runtime
+    * Unknown operator versions need to throw an error
+    * The operator version and upgraders are built into the runtime for BC.
+    * Allow for the addition of optional keyword-only arguments without a version bump or FC concern
+    * Since additional operators can be introduced in upgraders, tracing based selective build should also cover upgraders: easier for BC because the new runtimes goes with the upgraders. 
```

Comment:
We should also consider the timeline for mobile to no longer use upgraders by requiring models that are too old update themselves before deployment